### ISSUE TITLE
Odd error output when vulnerabilities found in package

### ISSUE THREAD

**Issue by pdehaan** [nodesecurity/gulp-nsp#12]
Here's my gulpfile:

``` js
var gulp = require('gulp');
var gulpNSP = require('gulp-nsp');

// To check your project
gulp.task('nsp', function (done) {
  gulpNSP({
    // output: 'summary',
    package: require('./package.json')
  }, done);
});
```

... and here's my package.json (note: **tar@1** has knonwn vulnerabilities, per https://nodesecurity.io/advisories):

``` json
{
  "name": "gulp-nsp-test",
  "version": "1.0.0",
  "author": "Peter deHaan <peter@deseloper.com> (http://nodeexamples.com/)",
  "dependencies": {
    "tar": "1.0.3"
  },
  "devDependencies": {
    "gulp": "3.9.0",
    "gulp-nsp": "2.1.0"
  },
  "license": "WTFPL",
  "main": "index.js"
}
```

And here's my `$ gulp nsp` output:

``` sh
$ gulp nsp

[19:19:01] Using gulpfile ~/dev/gulp-nsp-test/gulpfile.js
[19:19:01] Starting 'nsp'...
[19:19:02] 'nsp' errored after 542 ms
[19:19:02] Error: (+) 1 vulnerabilities found
┌───────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│               │ Symlink Arbitrary File Overwrite                                                                                     │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Name          │ tar                                                                                                                  │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Installed     │ 1.0.3                                                                                                                │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Vulnerable    │ <2.0.0                                                                                                               │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Patched       │ >=2.0.0                                                                                                              │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Path          │ gulp-nsp-test > tar                                                                                                  │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ More Info     │ https://nodesecurity.io/advisories/57                                                                                │
└───────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

    at formatError (/Users/pdehaan/.npm-packages/lib/node_modules/gulp/bin/gulp.js:169:10)
    at Gulp.<anonymous> (/Users/pdehaan/.npm-packages/lib/node_modules/gulp/bin/gulp.js:195:15)
    at Gulp.emit (events.js:107:17)
    at Gulp.Orchestrator._emitTaskDone (/Users/pdehaan/dev/gulp-nsp-test/node_modules/gulp/node_modules/orchestrator/index.js:264:8)
    at /Users/pdehaan/dev/gulp-nsp-test/node_modules/gulp/node_modules/orchestrator/index.js:275:23
    at finish (/Users/pdehaan/dev/gulp-nsp-test/node_modules/gulp/node_modules/orchestrator/lib/runTask.js:21:8)
    at cb (/Users/pdehaan/dev/gulp-nsp-test/node_modules/gulp/node_modules/orchestrator/lib/runTask.js:29:3)
    at /Users/pdehaan/dev/gulp-nsp-test/node_modules/gulp-nsp/index.js:37:14
    at /Users/pdehaan/dev/gulp-nsp-test/node_modules/gulp-nsp/node_modules/nsp/lib/check.js:214:7
    at /Users/pdehaan/dev/gulp-nsp-test/node_modules/gulp-nsp/node_modules/nsp/node_modules/wreck/lib/index.js:472:20
```

**Comment by pdehaan** [comment]
I can set `stopOnError: false` in my Gulp task, but I don't wanna... I want the `nsp` task to fail the build, just not spit the error stack into das console.

```
 gulp nsp
[19:27:29] Using gulpfile ~/dev/gulp-nsp-test/gulpfile.js
[19:27:29] Starting 'nsp'...
[19:27:30] (+) 1 vulnerabilities found
┌───────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│               │ Symlink Arbitrary File Overwrite                                                                                     │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Name          │ tar                                                                                                                  │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Installed     │ 1.0.3                                                                                                                │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Vulnerable    │ <2.0.0                                                                                                               │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Patched       │ >=2.0.0                                                                                                              │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Path          │ gulp-nsp-test > tar                                                                                                  │
├───────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ More Info     │ https://nodesecurity.io/advisories/57                                                                                │
└───────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

[19:27:30] Finished 'nsp' after 1.24 s


$ echo $?
0
```

**Comment by joeycozza** [comment]
Hey @pdehaan I noticed that as well, but didn't think too much of it.  I'll dig into this and see if we can have a more clean output during the failure.  

I'm not quite sure, but I think it might be a gulp specific thing since we pass an error to gulp's "done" callback. Anyways, I'll look into it and get back to you/fix it if I can.

**Comment by joeycozza** [comment]
I did a bit of a deep dive on this issue, and it appears that when we call gulp's done function with the error output string, Orchestrator does some things with timing and wrapping of the string into an object.  Then this gulp event is called.

```
  gulpInst.on('task_err', function(e) {
    var msg = formatError(e);
    var time = prettyTime(e.hrDuration);
    gutil.log(
      '\'' + chalk.cyan(e.task) + '\'',
      chalk.red('errored after'),
      chalk.magenta(time)
    );
    gutil.log(msg);
  });
```

At that point, e was automatically wrapped by Orchestrator.

```
// Format orchestrator errors
function formatError(e) {

  if (!e.err) {
    return e.message;
  }

  // PluginError
  if (typeof e.err.showStack === 'boolean') {
    return e.err.toString();
  }

  // Normal error
  if (e.err.stack) {
    return e.err.stack;
  }

  // Unknown (string, number, etc.)
  return new Error(String(e.err)).stack;
}
```

The formatError function then gets called, and no matter what I did to the string being passed into gulp's done function (wrapping it in an object, etc), it was always wrapped again by Orchestrator and the final line of formatError was run.  It always printed off that stack.  

Trying to dive into the Orchestrator proved unfruitful for me. It was WAY too cryptic.  Sorry I couldn't help you out with that issue. 

We pass the output into gulp's done function on line 42 of gulp-nsp.  If there is a way, I believe that is the place to look at. If somehow we could get the `e` in formatError to not have `.err` attached on it, then it would just print off `e.message`, which seems like what you'd want. If you'd like to look further, that would be great, but I think I'm done trying to figure this out. Sorry I couldn't help.

**Comment by pdehaan** [comment]
@joeycozza I haven't done a lot of testing on this yet, but I think wrapping it in a gutil.PluginError may work:

``` js
'use strict';

var gutil = require('gulp-util');
var Nsp = require('nsp');

var PLUGIN_NAME = require('./package.json').name;

var rsGulp = function (params, callback) {
  var payload = {};
  var formatter = Nsp.formatters.default;

  if (params.package) {
    payload.package = params.package;
  }

  if (params.shrinkwrap) {
    payload.shrinkwrap = params.shrinkwrap;
  }

  if (params.output) {
    formatter = Nsp.formatters[params.output];
  }

  Nsp.check(payload, function (err, data) {
    var output = formatter(err, data);
    var pluginErr = new gutil.PluginError(PLUGIN_NAME, output);

    if (err) {
      return callback(pluginErr);
    }

    if (data.length > 0) {
      return callback(pluginErr);
    }

    if (params.stopOnError === false || data.length === 0) {
      gutil.log(output);
      return callback();
    }
  });
};

module.exports = rsGulp;
```

---

Here's my test package.json and output:

``` json
{
  "name": "gulp-nsp-test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "Peter deHaan <peter@deseloper.com> (http://nodeexamples.com/)",
  "license": "WTFPL",
  "dependencies": {
    "tar": "1.0.3"
  },
  "devDependencies": {
    "gulp": "3.9.0",
    "gulp-nsp": "2.1.0"
  }
}
```

``` sh
$ gulp nsp
[01:12:38] Using gulpfile ~/dev/tmp/del/gulp-nsp-test/gulpfile.js
[01:12:38] Starting 'nsp'...
[01:12:40] 'nsp' errored after 1.84 s
[01:12:40] Error in plugin 'gulp-nsp'
Message:
    (+) 1 vulnerabilities found
┌───────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│               │ Symlink Arbitrary File Overwrite                                                                        │
├───────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Name          │ tar                                                                                                     │
├───────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Installed     │ 1.0.3                                                                                                   │
├───────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Vulnerable    │ <2.0.0                                                                                                  │
├───────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Patched       │ >=2.0.0                                                                                                 │
├───────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Path          │ gulp-nsp-test > tar                                                                                     │
├───────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ More Info     │ https://nodesecurity.io/advisories/57                                                                   │
└───────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────┘


$ echo $?
1
```

**Comment by joeycozza** [comment]
Looks like you figured it out @pdehaan . Nice work.

**Comment by nlf** [comment]
closed via #18 and published in 2.3.0
